src/kubernetes.cc

```diff
@@ -792,30 +792,39 @@ void KubernetesReader::WatchMaster(
   http::client::request request(endpoint);
   request << boost::network::header(
       "Authorization", "Bearer " + KubernetesApiToken());
-  const bool verbose = config_.VerboseLogging();
-  if (verbose) {
-    LOG(INFO) << "WatchMaster(" << name << "): Contacting " << endpoint;
-  }
   try {
+    const bool verbose = config_.VerboseLogging();
     if (verbose) {
-      LOG(INFO) << "Locking completion mutex";
+      LOG(INFO) << "WatchMaster(" << name << "): Will retry "
+                << config_.KubernetesUpdaterWatchRetries() << " times";
     }
-    // A notification for watch completion.
-    std::mutex completion_mutex;
-    std::unique_lock<std::mutex> watch_completion(completion_mutex);
-    Watcher watcher(
-      endpoint,
-      [=](json::value raw_watch) {
-        WatchEventCallback(callback, name, std::move(raw_watch));
-      },
-      std::move(watch_completion), verbose);
-    http::client::response response = client.get(request, std::ref(watcher));
-    if (verbose) {
-      LOG(INFO) << "Waiting for completion";
+    for (int i = 0; i < config_.KubernetesUpdaterWatchRetries(); i++) {
+      if (verbose) {
+        LOG(INFO) << "WatchMaster(" << name << "): Contacting " << endpoint
+                  << " attempt #" << i;
+      }
+      // A notification for watch completion.
+      std::mutex completion_mutex;
+      std::unique_lock<std::mutex> watch_completion(completion_mutex);
+      Watcher watcher(
+        endpoint,
+        [=](json::value raw_watch) {
+          WatchEventCallback(callback, name, std::move(raw_watch));
+        },
+        std::move(watch_completion), verbose);
+      http::client::response response = client.get(request, std::ref(watcher));
+      if (verbose) {
+        LOG(INFO) << "Waiting for completion attempt #" << i;
+      }
+      std::lock_guard<std::mutex> await_completion(completion_mutex);
+      if (verbose) {
+        LOG(INFO) << "WatchMaster(" << name << ") completed " << body(response)
+                  << "attempt #" << i;
```

Comment:
Same here, I'd print the attempt only if `i` is non-zero.

Reply:
Done.